### DISCUSSION TITLE
Windows Terminal Ubuntu-20.04 | Fatal: can't open lock file /run/xtables.lock: Permission denied

### DISCUSSION THREAD

**Discussion by newcanopies** [microsoft/WSL#6780] [Q&A]
# Environment

Microsoft Windows [Version 10.0.19041.867]
Ubuntu Focal 20.04
Linux   version 5.4.72-microsoft-standard-WSL2 (oe-user@oe-host) (gcc version 8.2.0   (GCC)) #1 SMP Wed Oct 28 23:40:43 UTC 2020

# Steps to reproduce

opening a Windows Terminal Ubuntu shell outputs following:

```
sysctl: permission denied on key "net.ipv4.ip_forward"
sysctl: permission denied on key "net.ipv4.conf.all.route_localnet"
Fatal: can't open lock file /run/xtables.lock: Permission denied
Fatal: can't open lock file /run/xtables.lock: Permission denied
```

What is the cause?
Thank you

**Comment by therealkenc** [comment]
Likely something has been inserted into one of your startup scripts (`.bashrc` or similar) which is calling `sysctl(8)` as an unprivileged user ('you'), and you do not have permission to modify `net.ipv4.ip_forward` etc. A big hammer fix would be to install a differently-named distribution [Ubuntu-20.04](https://www.microsoft.com/en-us/p/ubuntu-2004-lts/9n6svws3rx71) vs [Ubuntu](https://www.microsoft.com/en-us/p/ubuntu/9nblggh4msv6), or vise-versa, start it with `wsl.exe -d <distro name>` and confirm whether the unadulterated install starts without error.

**Reply by HomyeeKing** [reply]
hey @therealkenc  , can you help me with this https://github.com/microsoft/WSL/discussions/6835?
I met some `command not found ` issue

**Comment by newcanopies** [comment]
thank you @therealkenc 

errors as below from running various programs, despite `$ export   XDG_RUNTIME_DIR=/run/user/1000`
![image](https://user-images.githubusercontent.com/65022111/114077174-cb239800-98a7-11eb-8386-49f7a1860155.png)

**Comment by newcanopies** [comment]
@therealkenc 

gedit error inside Visual Studio Code
![image](https://user-images.githubusercontent.com/65022111/114077686-5a30b000-98a8-11eb-8ce0-c8bfabc57a8f.png)

**Reply by therealkenc** [reply]
> are any of these Windows/cmd/PowerShell based errors,

No, they're all from Linux-land.

>  the fresh Ubuntu install should wipe the slate clean on all of them?

Yes. The errors are from configuration changes made since a fresh install.

> Fatal: can't open lock file /run/xtables.lock: Permission denied

User 'you' does not have permission to that file.

> gedit error inside Visual Studio Code

The `code.exe` and `gedit` apps are two different programs. No idea what is going on there. VS Code works from a fresh install (all things being equal).

> running w, who -H, who -u returns blank!

That's three digit #573

**Reply by therealkenc** [reply]
> Fatal: can't open lock file /run/xtables.lock: Permission denied

Find the command that is attempting to access that file. In other words, the source of the error. Or install a new distribution.

> Gedit has the same issue in Windows Terminal@Ubuntu

Windows Terminal is not aware of `gedit`, and unrelated to Ubuntu.

**Reply by therealkenc** [reply]
> 172.xx.xxx.1: command not found

You did a `sudo 172.xx.xxx.1`, which is not a command.

> Back on WSL1 neither VSCode nor WindowsTerminal produced the below gedit error, is it WSL2 specific ?

Windows Terminal is not aware of gedit, and unrelated to Ubuntu. No idea what is going on there. VS Code works from a fresh install on WSL2 (all things being equal).

**Reply by therealkenc** [reply]
> hope this is more clear of the actual command:

You had posted the actual command previously.

```
$ sudo $(cat /etc/resolv.conf | grep nameserver | awk '{ print $2 }')
```

Expands to:
```
$ sudo 172.28.186.1
```

And `172.28.186.1` is not a command.

Have you installed [Win-KeX](https://www.kali.org/docs/wsl/win-kex/)? If you have, you will get better eyeballs in the [Kali Forums](https://forums.kali.org/).

**Reply by therealkenc** [reply]
That line is in the correct format.

**Reply by therealkenc** [reply]
> returns
> Fatal: can't open lock file /run/xtables.lock: Permission denied

No, [`cat(1)`](https://man7.org/linux/man-pages/man1/cat.1.html), [`grep(1)`](https://man7.org/linux/man-pages/man1/grep.1.html), and [`awk(1)`](https://man7.org/linux/man-pages/man1/gawk.1.html) do not access `/run/xtables.lock`. You can try it from a `bash` prompt:

![image](https://user-images.githubusercontent.com/7049700/115771898-56e6fb00-a363-11eb-8b2b-dd6f274a2113.png)

> In conclusion, my WSL2 on Windows 19041, has a bug with User

You `permission denied` error on `/run/xtables.lock` is unrelated to #573.

> How likely is it to be resolved with a new Distribution?

None of these lines will appear in an unadulterated `.bashrc`, ie a fresh install.

What software package has added these lines to your `.bashrc`?

**Reply by therealkenc** [reply]
> If whoami does return my username,
> should w return 0 users (which is the case) in my WSL2 bash, or show my username as logged in?

[`whoami(1)`](https://man7.org/linux/man-pages/man1/whoami.1.html) reads `/etc/passwd`. `who(1)` (aka `w`) reads `/var/run/utmp`. They are different commands that do different things.

> login(1) myusername

```
$ sudo login -f yourusername
```

You `permission denied error` on `/run/xtables.lock` is unrelated to #573.
